libafl_nyx/src/helper.rs

```diff
@@ -0,0 +1,130 @@
+/// [`NyxHelper`] is used to wrap `NyxProcess`
+use std::{
+    fmt::{self, Debug},
+    path::Path,
+    time::Duration,
+};
+
+use libafl::Error;
+use libnyx::{NyxProcess, NyxReturnValue};
+
+pub struct NyxHelper {
+    pub nyx_process: NyxProcess,
+    /// real size of trace_bits
+    pub real_map_size: usize,
+    // real size of the trace_bits
+    pub map_size: usize,
+    /// shared memory with instruction bitmaps
+    pub trace_bits: *mut u8,
+}
+
+const MAX_FILE: u32 = 1024 * 1024;
+#[derive(Clone, Copy, Debug)]
+pub enum NyxProcessType {
+    /// stand alone mode
+    ALONE,
+    /// parallel mode's parent, used to create snapshot
+    PARENT,
+    /// parallel mode's child, consume snapshot and execute
+    CHILD,
+}
+impl NyxHelper {
+    /// create `NyxProcess` and do basic settings
+    /// It will convert instance to parent or child using `parent_cpu_id` when set`parallel_mode`
+    pub fn new(
+        target_dir: &Path,
+        cpu_id: u32,
+        snap_mode: bool,
+        parallel_mode: bool,
+        parent_cpu_id: Option<u32>,
+    ) -> Result<Self, Error> {
+        let sharedir = target_dir
+            .to_str()
+            .expect("unable to convert target_dir to str");
```

Comment:
maybe use ? not expect() here

Reply:
the type of `to_str()` is `Option`, so I can't use `?`

Reply:
You can use something like `.unwrap_or_else(||Error::ilegal_argument(..`  or simply an if, here

Reply:
Either is ok
We want to avoid panicking inside our lib

Reply:
emmm... I just keep it unmodified. The doc says `.to_str() Yields a &str if the is valid unicode`, it's more like the misuse of std library instead of our API if it returns `none`. So there's no need to handle it in our crate.

Reply:
OK

Reply:
The docs also say paths containing non-unicode characters is "perfectly legal", so, in theory, either we want to forward the bytes to nyx directly, here, or use `to_string_lossy`.
Since we only do Linux, it's probably fine in this case, but keep it in mind in general